File: Go/Rotate Array.go
// 4ms, 95.76%; 3.2MB, 100.00%
func rotate(nums []int, k int) {
	k %= len(nums)
	if k == 0 {
		return
	}
	beg, last := 0, nums[0]
	for i, j := k, 1; j < len(nums); i, j = (i+k)%len(nums), j+1 {
		last, nums[i] = nums[i], last
		if i == beg {
			i, beg = i+1, beg+1
			last = nums[i]
		}
	}
	nums[beg] = last
}

File: Java/Rotate Array.java
// 0ms, 100.00%; 39.8MB, 5.77%
class Solution {
    public static void rotate(int[] nums, int k) {
        if ((k %= nums.length) == 0) {
            return;
        }
        int beg = 0, last = nums[0];
        for (int i = k, j = 1; i != beg || j != nums.length; i = (i + k) % nums.length, ++j) {
            int tmp = nums[i];
            nums[i] = last;
            last = tmp;
            if (i == beg) {
                last = nums[(i = ++beg)];
            }
        }
        nums[beg] = last;
    }
}
